R/contrs.R

```diff
@@ -0,0 +1,84 @@
+#' Deviation Contrast Matrix
+#'
+#' Build a deviation contrast matrix, a type of _effects contrast_ matrix.
+#'
+#' @inheritParams stats::contr.sum
+#'
+#' @details
+#' In effects coding, unlike dummy coding ([stats::contr.treatment()]), each
+#' contrasts sums to 0. In regressions models, this results in an intercept that
+#' represents the (unweighted) average of the group means. In ANOVA settings,
+#' this also guarantees that lower order effects represent _main_ effects (and
+#' not _simple_ or _conditional_ effects, as is the case when using
+#' [stats::contr.treatment()], which is `R`'s default).
+#' \cr\cr
+#' `contr.deviation`, unlike [stats::contr.sum()], also has the added benefit
+#' that the factor-related coefficients are interpretable. In fact, they
+#' represent the same contrasts as those of [stats::contr.treatment()]: the
+#' difference of each level from the base level.
+#'
+#' @seealso [stats::contr.sum()]
+#'
+#' @examples
+#' \dontrun{
+#' data("mtcars")
+#'
+#' mtcars$cyl <- factor(mtcars$cyl)
+#'
+#' c.treatment <- cbind(Intercept = 1, contrasts(mtcars$cyl))
+#' solve(c.treatment)
+#' #>            4 6 8
+#' #> Intercept  1 0 0  # mean of the 1st level
+#' #> 6         -1 1 0  # 2nd level - 1st level
+#' #> 8         -1 0 1  # 3rd level - 1st level
+#'
+#' contrasts(mtcars$cyl) <- contr.sum
+#' c.sum <- cbind(Intercept = 1, contrasts(mtcars$cyl))
+#' solve(c.sum)
+#' #>                4      6      8
+#' #> Intercept  0.333  0.333  0.333   # overall mean
+#' #>            0.667 -0.333 -0.333   # 2/3 * ({2nd, 3rd} - 1st)
+#' #>           -0.333  0.667 -0.333   # 2/3 * ({1st, 3rd} - 2nd)
+#'
+#'
+#' contrasts(mtcars$cyl) <- contr.deviation
+#' c.deviation <- cbind(Intercept = 1, contrasts(mtcars$cyl))
+#' solve(c.deviation)
+#' #>                4     6     8
+#' #> Intercept  0.333 0.333 0.333   # overall mean
+#' #> 6         -1.000 1.000 0.000   # 2nd level - 1st level
+#' #> 8         -1.000 0.000 1.000   # 3rd level - 1st level
+#'
+#' ## With Interactions -----------------------------------------
+#' mtcars$am <- factor(mtcars$am)
+#' contrasts(mtcars$am) <- contr.deviation
+#'
+#' mtcars <- mtcars[order(mtcars$cyl, mtcars$am),]
```

Comment:
you could use `datawizard` syntax here with `data_arrange()`

Reply:
sure